DRA E2E: promote one test to conformance

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

All of these tests already ran before for a while in
        https://testgrid.k8s.io/sig-release-master-informing#kind-master-beta&include-filter-by-regex=DRA
    
Now that DRA is GA, they also ran e.g. in
        https://testgrid.k8s.io/sig-release-master-blocking#kind-master&include-filter-by-regex=DRA&include-filter-by-regex=DRA
        https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&include-filter-by-regex=DRA
    
For the sake of starting conservatively, only a single basic test which exercises ResourceSlice, DeviceClass, ResourceClaimTemplate, ResourceClaim and ResourceClaim/Status gets promoted to a conformance test:
    
        dra/dra.go:747: [sig-node] [DRA] control plane supports claim and class parameters [Conformance]
    

Documentation for other tests which could get promoted to conformance gets  added for future reference. Such documentation is useful also when the tests  never become conformance tests. Whether they are worth turning into conformance tests remains to be seen, in particular if that means duplicating  and potentially rewriting the test body as it was done for the one test which  gets promoted.

#### Which issue(s) this PR is related to:

https://github.com/kubernetes/kubernetes/issues/114183 (not 100% coverage!)
KEP: https://github.com/kubernetes/enhancements/issues/4381

#### Special notes for your reviewer:

/area conformance
cc @kubernetes/sig-architecture-pr-reviews @kubernetes/sig-node-pr-reviews @kubernetes/cncf-conformance-wg

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
DRA E2E: promote one test to conformance

All of these tests already ran before for a while in
    https://testgrid.k8s.io/sig-release-master-informing#kind-master-beta&include-filter-by-regex=DRA

Now that DRA is GA, they also ran e.g. in
    https://testgrid.k8s.io/sig-release-master-blocking#kind-master&include-filter-by-regex=DRA&include-filter-by-regex=DRA
    https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&include-filter-by-regex=DRA

For the sake of starting conservatively, only a single basic test which
exercises ResourceSlice, ResourceClaimTemplate, ResourceClaim and
ResourceClaim/Status gets promoted to a conformance test:

    dra/dra.go:747: [sig-node] [DRA] control plane supports claim and class parameters [Conformance]

Documentation for other tests which could get promoted to conformance gets
added for future reference. Such documentation is useful also when the tests
never become conformance tests. Whether they are worth turning into
conformance tests remains to be seen, in particular if that means duplicating
and potentially rewriting the test body as it was done for the one test which
gets promoted.

pohly commented on test/e2e/dra/dra.go:
```diff
@@ -921,74 +1009,6 @@ var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), framework.With
 			}).WithTimeout(f.Timeouts.PodDelete).Should(gomega.HaveField("Status.Allocation", (*resourceapi.AllocationResult)(nil)))
 		})
 
-		f.It("must be possible for the driver to update the ResourceClaim.Status.Devices once allocated", f.WithFeatureGate(features.DRAResourceClaimDeviceStatus), func(ctx context.Context) {
```
This shouldn't have been in `singleNodeTests` because then it runs with and without kubelet. It gets moved to a better location such that it only runs once overall. I can also merge this separately.

BenTheElder commented:
> All of these tests already ran before for a while (possibly under slightly different names or labels) in:
https://testgrid.k8s.io/sig-release-master-informing#kind-master-beta&include-filter-by-regex=DRA

Can we clarify that? It's usually on the PR author to provide the evidence of mapping the proposed tests to stable test results.

Some of the tests in that link, which are tagged "ConformanceCandidate", are being skipped and not passing because they require 3+ schedule-able nodes (which would also be new for a conformance test, currently we require 2).

BenTheElder commented:
Conformance also doesn't currently require that the control plane even exists as a node in any form, it seems like noise to tag most of these tests with "control plane with single node" in the name. Or else if these tests depend on there only being a single control plane node in some way, we should rewrite them to not depend on that sort of detail.

Clusters should be equally conformant with HA control planes and single instance control planes and the exact details of how the apiserver is hosted are not important, just the behaviors of the API.

aojea commented:
I do not feel this has to go to conformance without more soak time of the code in GA, these tests were effectively tested with beta until today

pohly commented:
> I think there are 2 AI that blocks the Conformance promotion
> - the single/multinode thing

Resolved in https://github.com/kubernetes/kubernetes/pull/133141.

For the updated `test/conformance/testdata/conformance.yaml` please see https://github.com/kubernetes/kubernetes/pull/132554, my GA tracking branch/PR.

johnbelamaric commented:
> * wait a release (@aojea)

Not an option, all GA, non-optional features must have conformance tests.

> * just enough to hit the new endpoints (ResourceClaim, ResourceClaim/Status, ResourceClaimTemplate, ResourceSlice) - that could be a single test.

Let's go with the minimal set for now, we can always layer in more later. This reduces risk for distros. What about DeviceClass?

pohly commented:
/test pull-kubernetes-integration

Unrelated flake.

pohly commented:
/assign @johnbelamaric

stlaz commented:
The test appears to be very flaky during pull request runs, apparently the pod that's getting created tends to be unschedulable.
@pohly could you have a look?

Failed run links from the past 24 hrs:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/131260/pull-kubernetes-e2e-gce/1952515411312381952
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133373/pull-kubernetes-e2e-gce/1952421018924486656
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/132812/pull-kubernetes-e2e-gce/1952320607337058304

yliaog commented:
The single test to promote to conformance below does not seem to be flaky.
    dra/dra.go:747: [sig-node] [DRA] control plane supports claim and class parameters [Conformance]

The following test seems to be flaky, it is ConformanceCandidate, its flakiness needs to be fixed before promotion.
Kubernetes e2e suite.[It] [sig-node] [DRA] control plane [ConformanceCandidate] supports reusing resources

yliaog commented:
@stlaz @BenTheElder could you please double check if the following test can be promoted to conformance?

Kubernetes e2e suite.[It] [sig-node] [DRA] control plane [ConformanceCandidate] supports claim and class parameters

BenTheElder commented:
"supports claim and class" only had one flake on GCE in recent history, it is flaking somewhat on a CAPZ job I don't recognize and am doubtful we should block on.

https://storage.googleapis.com/k8s-triage/index.html?test=supports%20claim%20and%20class

yliaog commented:
I also investigated the cause for the flakiness of the test:
Kubernetes e2e suite.[It] [sig-node] [DRA] control plane [ConformanceCandidate] supports reusing resources

the test creates 10 pods at the same time, all of the 10 pods reuse the same device, which means each pod needs to be scheduled, then deleted, sequentially, to be able to reuse the same device. The sequential scheduling and deletion of pod takes time, currently, the wait timeout for all of the 10 pods to be scheduled is 80 seconds, which in some cases are not sufficient, as shown in the following log.

At 2025-08-05 20:01:43, the first try of scheduling the pod has started, at 2025-08-05 20:03:07, the wait timed out. during this period, some pods have been scheduled, and deleted, but not all of the 10 pods. so increasing the wait timeout, or reducing the # of pods, should fix the test flakiness.


https://prow.k8s.io/13b8c7b4-1951-4693-9393-83b4a3068d7d

I0805 20:03:13.208285 9189 dump.go:53] At 2025-08-05 20:01:43 +0000 UTC - event for tester-10: {resource_claim } FailedResourceClaimCreation: PodResourceClaim my-inline-claim: resource claim template "tester-10": resourceclaimtemplate.resource.k8s.io "tester-10" not found
I0805 20:03:13.208289 9189 dump.go:53] At 2025-08-05 20:01:50 +0000 UTC - event for tester-1: {default-scheduler } FailedScheduling: 0/4 nodes are available: 1 node(s) had untolerated taint {node-role.kubernetes.io/control-plane: }, 1 timed out trying to allocate devices, 2 cannot allocate all claims. still not schedulable, preemption: 0/4 nodes are available: 4 Preemption is not helpful for scheduling.
I0805 20:03:13.208292 9189 dump.go:53] At 2025-08-05 20:01:59 +0000 UTC - event for tester-2: {default-scheduler } Scheduled: Successfully assigned dra-1246/tester-2 to bootstrap-e2e-minion-group-rk4q
I0805 20:03:13.208295 9189 dump.go:53] At 2025-08-05 20:01:59 +0000 UTC - event for tester-2: {kubelet bootstrap-e2e-minion-group-rk4q} FailedPrepareDynamicResources: Failed to prepare dynamic resources: prepare dynamic resources: DRA driver dra-1246.k8s.io is not registered
I0805 20:03:13.208297 9189 dump.go:53] At 2025-08-05 20:02:07 +0000 UTC - event for tester-3: {default-scheduler } FailedScheduling: 0/4 nodes are available: 1 node(s) had untolerated taint {node-role.kubernetes.io/control-plane: }, 3 cannot allocate all claims. still not schedulable, preemption: 0/4 nodes are available: 4 Preemption is not helpful for scheduling.
I0805 20:03:13.208300 9189 dump.go:53] At 2025-08-05 20:02:13 +0000 UTC - event for tester-4: {default-scheduler } Scheduled: Successfully assigned dra-1246/tester-4 to bootstrap-e2e-minion-group-rk4q
I0805 20:03:13.208311 9189 dump.go:53] At 2025-08-05 20:02:18 +0000 UTC - event for tester-5: {default-scheduler } FailedScheduling: 0/4 nodes are available: 1 node(s) had untolerated taint {node-role.kubernetes.io/control-plane: }, 3 cannot allocate all claims. still not schedulable, preemption: 0/4 nodes are available: 4 Preemption is not helpful for scheduling.
I0805 20:03:13.208314 9189 dump.go:53] At 2025-08-05 20:02:25 +0000 UTC - event for tester-6: {default-scheduler } Scheduled: Successfully assigned dra-1246/tester-6 to bootstrap-e2e-minion-group-rk4q
I0805 20:03:13.208317 9189 dump.go:53] At 2025-08-05 20:02:26 +0000 UTC - event for tester-6: {kubelet bootstrap-e2e-minion-group-rk4q} FailedPrepareDynamicResources: Failed to prepare dynamic resources: prepare dynamic resources: DRA driver dra-1246.k8s.io is not registered
I0805 20:03:13.208321 9189 dump.go:53] At 2025-08-05 20:02:30 +0000 UTC - event for tester-7: {default-scheduler } FailedScheduling: 0/4 nodes are available: 1 node(s) had untolerated taint {node-role.kubernetes.io/control-plane: }, 3 cannot allocate all claims. still not schedulable, preemption: 0/4 nodes are available: 4 Preemption is not helpful for scheduling.
I0805 20:03:13.208326 9189 dump.go:53] At 2025-08-05 20:02:35 +0000 UTC - event for tester-8: {default-scheduler } Scheduled: Successfully assigned dra-1246/tester-8 to bootstrap-e2e-minion-group-rk4q
I0805 20:03:13.208329 9189 dump.go:53] At 2025-08-05 20:02:36 +0000 UTC - event for tester-8: {kubelet bootstrap-e2e-minion-group-rk4q} FailedPrepareDynamicResources: Failed to prepare dynamic resources: prepare dynamic resources: DRA driver dra-1246.k8s.io is not registered
I0805 20:03:13.208332 9189 dump.go:53] At 2025-08-05 20:02:39 +0000 UTC - event for tester-9: {default-scheduler } FailedScheduling: 0/4 nodes are available: 1 node(s) had untolerated taint {node-role.kubernetes.io/control-plane: }, 3 cannot allocate all claims. still not schedulable, preemption: 0/4 nodes are available: 4 Preemption is not helpful for scheduling.
I0805 20:03:13.208414 9189 dump.go:53] At 2025-08-05 20:02:42 +0000 UTC - event for tester-10: {default-scheduler } FailedScheduling: 0/4 nodes are available: 1 node(s) had untolerated taint {node-role.kubernetes.io/control-plane: }, 3 cannot allocate all claims. still not schedulable, preemption: 0/4 nodes are available: 4 Preemption is not helpful for scheduling.
I0805 20:03:13.208421 9189 dump.go:53] At 2025-08-05 20:02:46 +0000 UTC - event for tester-1: {default-scheduler } Scheduled: Successfully assigned dra-1246/tester-1 to bootstrap-e2e-minion-group-rk4q
I0805 20:03:13.208424 9189 dump.go:53] At 2025-08-05 20:02:47 +0000 UTC - event for tester-1: {kubelet bootstrap-e2e-minion-group-rk4q} FailedPrepareDynamicResources: Failed to prepare dynamic resources: prepare dynamic resources: DRA driver dra-1246.k8s.io is not registered
I0805 20:03:13.208427 9189 dump.go:53] At 2025-08-05 20:02:53 +0000 UTC - event for tester-7: {default-scheduler } Scheduled: Successfully assigned dra-1246/tester-7 to bootstrap-e2e-minion-group-rk4q
I0805 20:03:13.208429 9189 dump.go:53] At 2025-08-05 20:02:54 +0000 UTC - event for tester-7: {kubelet bootstrap-e2e-minion-group-rk4q} FailedPrepareDynamicResources: Failed to prepare dynamic resources: prepare dynamic resources: DRA driver dra-1246.k8s.io is not registered
I0805 20:03:13.208432 9189 dump.go:53] At 2025-08-05 20:03:00 +0000 UTC - event for tester-5: {default-scheduler } Scheduled: Successfully assigned dra-1246/tester-5 to bootstrap-e2e-minion-group-rk4q
I0805 20:03:13.208436 9189 dump.go:53] At 2025-08-05 20:03:01 +0000 UTC - event for tester-5: {kubelet bootstrap-e2e-minion-group-rk4q} FailedPrepareDynamicResources: Failed to prepare dynamic resources: prepare dynamic resources: DRA driver dra-1246.k8s.io is not registered
I0805 20:03:13.208439 9189 dump.go:53] At 2025-08-05 20:03:06 +0000 UTC - event for tester-9: {default-scheduler } FailedScheduling: skip schedule deleting pod: dra-1246/tester-9
I0805 20:03:13.208442 9189 dump.go:53] At 2025-08-05 20:03:07 +0000 UTC - event for tester-3: {default-scheduler } FailedScheduling: running Bind plugin "DefaultBinder": pods "tester-3" not found
I0805 20:03:13.208445 9189 dump.go:53] At 2025-08-05 20:03:09 +0000 UTC - event for dra-test-driver: {replicaset-controller } SuccessfulDelete: Deleted pod: dra-test-driver-2m8p9
I0805 20:03:13.208448 9189 dump.go:53] At 2025-08-05 20:03:09 +0000 UTC - event for dra-test-driver-2m8p9: {kubelet bootstrap-e2e-minion-group-rk4q} Killing: Stopping container pause

johnbelamaric commented:
> the test creates 10 pods at the same time, all of the 10 pods reuse the same device, which means each pod needs to be scheduled, then deleted, sequentially, to be able to reuse the same device. The sequential scheduling and deletion of pod takes time, currently, the wait timeout for all of the 10 pods to be scheduled is 80 seconds, which in some cases are not sufficient, as shown in the following log.

10 seems arbitrary, perhaps we reduce to 3 or 4, which will reduce the runtime as well.

BenTheElder commented on test/e2e/dra/dra.go:
```diff
@@ -1972,7 +2103,9 @@ var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), func() {
 		})
 
-		f.It("supports count/resourceclaims.resource.k8s.io ResourceQuota", framework.WithLabel("ConformanceCandidate") /* TODO: replace with framework.WithConformance() */, func(ctx context.Context) {
+		// This could be required for conformance, but there have been some rare known flakes in the quota controller, so it's not considered
```
Too bad :/

Do we have a tracking bug?

BenTheElder commented on test/conformance/testdata/conformance.yaml:
```diff
@@ -2843,6 +2843,16 @@
+    referencing the ResourceClaimTemplate. kube-scheduler must allocate a suitable
+    device from a ResourceSlice and copy the parameters into the allocation result.
+    kubelet must invoke a DRA driver such these parameters are active for the Pod
+    (not part of conformance).
```
If this is just documenting some expectation of how DRA should behave that is not tested, we should drop it from the description ...

If this is WRT the actual test behavior, we probably need to pick another test.

The test code is pretty heavily abstracted, it's not immediately apparent to my late night brain if this test depends on installing a driver / kubelet invoking a driver?

BenTheElder replied:
ACK.

BenTheElder reviewed (commented):
cc @dims re: alternative test suggestion^

yliaog commented:
@BenTheElder this PR picks the following test to promote, the picked test is not flaky.
Kubernetes e2e suite.[It] [sig-node] [DRA] control plane [ConformanceCandidate] supports claim and class parameters

the flaky test below is not promoted in this PR.
Kubernetes e2e suite.[It] [sig-node] [DRA] control plane [ConformanceCandidate] supports reusing resources

yliaog commented:
/retest

yliaog commented:
/retest

BenTheElder commented:
All cluster E2E is currently broken on master / 1.34 because apiserver panics on startup.

#133405

yliaog commented:
@BenTheElder @dims the PR has passed all tests, ready for review/approval

BenTheElder reviewed (commented):
IMHO: I have two TODOS that I think can be non-blocking but would be good to resolve:
1. Cleanup the test descriptions to not have conformance tests mention the details that are happening in the other variant of the same test that is not conformance. Just don't mention it, the mention is confusing.

2. Rewrite to avoid the local "framework" for duplicating the tests to conformance or not and "multi control plane" or not, it makes it harder to look at what this test is doing without local context and doesn't save much code.

In the 1.35 when the flakes have been resolved we can look at promoting at least also the other single test previously proposed.

WDYT @dims?

dims commented:
/test pull-audit-kind-conformance

dims commented:
@pohly please see https://gist.github.com/dims/1fc254d0e4d043ef18bbc84596785de2#file-endpoints-log and check if it looks right to you. (look for the endpoints that are present in the first table ... and absent in the second table)

dims commented:
/test pull-audit-kind-conformance

BenTheElder commented:
/test pull-kubernetes-kind-dra-n-1

pohly committed:
DRA E2E: clean up class creation for extended resource tests

The initial testing of DRAExtendedResource changed the common setup for all
tests so that each test runs with seven DeviceClasses. This is unnecessary and
confusing, for example when looking at the endpoint stats for the simple
conformance test ("Are the stats correct? The test should only create one
class.").

This update changes it so that each test which needs more than the default
class explicitly creates those itself. For clarity, the special -1 index gets a
name and naming of the DeviceClasses gets updated to keep the default class
simple (no additional suffix).

dims commented:
/test pull-audit-kind-conformance

dims commented:
/test pull-audit-kind-conformance

dims commented:
latest run of [pull-kubernetes-audit-kind-conformance](https://storage.googleapis.com/kubernetes-ci-logs/pr-logs/pull/133132/pull-kubernetes-audit-kind-conformance/1959592294910791680/build-log.txt) shows:

```
Extracting operations from audit files (filtering by swagger operations)...
SUMMARY
=======
Total Operations in Swagger:  1062
Operations in CI:             583
Operations in Pull:           592
Operations Added:             9
Operations Removed:           0
Net Change:                   +9

OPERATIONS ADDED IN PULL (NOT IN CI)
====================================
Count: 9

  1. createResourceV1DeviceClass
  2. createResourceV1NamespacedResourceClaim
  3. createResourceV1NamespacedResourceClaimTemplate
  4. createResourceV1ResourceSlice
  5. deleteResourceV1DeviceClass
  6. deleteResourceV1NamespacedResourceClaim
  7. readResourceV1NamespacedResourceClaim
  8. replaceResourceV1NamespacedResourceClaim
  9. replaceResourceV1NamespacedResourceClaimStatus

OPERATIONS REMOVED FROM PULL (IN CI BUT NOT PULL)
=================================================
Count: 0

No operations removed.

Analysis complete!
```

pohly commented:
So "OPERATIONS ADDED IN PULL (NOT IN CI)" means "operations covered only by new tests in the pull request"? More interesting would be "new operations that are not covered by tests".

The current output shows e.g. "createResourceV1DeviceClass" which is exercised by the new test. It does not show that "patchResourceV1DeviceClass" lacks coverage. If this was a GA promotion PR, then I'd like to know that.

dims commented:
@pohly yep i have that information and will surface it. need at least 2-3 more iterations to get there.

/test pull-kubernetes-audit-kind-conformance

dims commented:
/test pull-kubernetes-audit-kind-conformance

dims commented:
/test pull-kubernetes-audit-kind-conformance
